### PR TITLE
feat: update gradle dependencies

### PR DESCRIPTION
**Pull request by mobilekosmos**
- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/Parse-SDK-Android/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/Parse-SDK-Android/issues?q=is%3Aissue).

Related issue: #1159

- [ ] Add tests
- [ ] Add changes to documentation (guides, repository pages, in-code descriptions)
- [x] A changelog entry is created automatically using the pull request title (do not manually add a changelog entry)

### CODE REVIEW

**Review by L3K0V — APPROVED** [review]
Great job @mobilekosmos 

LGTM

**Comment by mtrezza** [comment]
I just noticed that CI is failing, do we need a separate PR to fix something or is this related to this PR?

**Comment by mobilekosmos** [comment]
It seems that all the mockito imports are not being resolved, could also see locally in my IDE, don't know why yet.

**Comment by mobilekosmos** [comment]
How to fix the spotless failing? Wouldn't it be possible to let the CI do the formatting automatically?

**Comment by mtrezza** [comment]
You should be able to run spotless to auto-correct what's possible locally and commit these changes.

**Comment by mobilekosmos** [comment]
"First-time contributors need a maintainer to approve running workflows."

**Review by L3K0V — APPROVED** [review]

**Comment by mtrezza** [comment]
Apologies for the delay, I started the CI.

**Comment by mtrezza** [comment]
@mobilekosmos Could you please take a look at the failing CI?

**Comment by mobilekosmos** [comment]
Despite such minor changes, the pain of a PR is big.

**Review by L3K0V — APPROVED** [review]

**Review by mtrezza — COMMENTED** [review]
This is upgrading the Facebook SDK from 12 to 13. Does this imply a breaking change for developers? If so we'd need to make this a major version increase.

**Comment by mobilekosmos** [comment]
> Does this imply a breaking change for developers? 
Shouldn't automated tests respond to this?

**Comment by mobilekosmos** [comment]
Btw. it seems that Facebook's Android SDK 13 urges to "migrate" to it: 
[](https://github.com/facebook/facebook-android-sdk/blob/main/CHANGELOG.md)
[](https://developers.facebook.com/blog/post/2022/02/24/introducing-facebook-platform-sdk-13-related-updates/?locale=en_US)

**Comment by mobilekosmos** [comment]
After seeing the amount of files changed with this PR I'm asking myself if it wouldn't be better, at least in the future, to accepto only one dependency update per each PR, more burocratic and painfull again, but maybe a better approach? Not sure yet.

**Comment by mtrezza** [comment]
@mobilekosmos Yes, I would not mix a major version upgrade of the FB SDK with other dependency upgrades, because it's an essential part of the SDK with its own complexities. Depending on what the breaking change in the FB SDK 13 is, we may have to declare a breaking change for the SDK if developers cannot simply upgrade the SDK without having to change their code.

**Comment by mobilekosmos** [comment]
The thing is if I understood it correctly apps using an old version of the facebook sdk will stop working anyways

**Comment by mtrezza** [comment]
As I understand it, only FB Login will not work, but could an app use other functionality of the FB SDK?

**Comment by mobilekosmos** [comment]
It would be a topic for another discussion, but the library should be modularized anyways, the main gradle Parse-SDK shouldn't contain ALL the modules it currently has, like google play services, etc. if you need facebook, twitter, etc. you must add parse-twitter separately and so on.
Regarding this PR, do what you want guys, I just wanted to help, but if it's so difficult I cannot invest more time.

**Comment by mtrezza** [comment]
We need to know whether this needs to be merged as a breaking change, that's all as far as I understand.

**Comment by mtrezza** [comment]
@L3K0V Could we get your opinion on whether upgrading the FB SDK by a major version should be considered a breaking change and we need to increment the major version of the Parse Android SDK as well?

When someone adds the Parse Android SDK, the FB SDK is added as a dependency well. Could they then just use any other feature of the FB SDK in their project, unrelated to the Parse SDK? If so, upgrading the Parse SDK may break their other FB related code?

**Comment by mtrezza** [comment]
Thanks, so we have to do a major version increment when we merge this.

**Comment by mtrezza** [comment]
I'm suggesting to fix https://github.com/parse-community/Parse-SDK-Android/issues/1158 first before we merge this, because any release without the migration logic is for most people likely a useless release.

**Comment by mtrezza** [comment]
Good news, #1158 has been merged, so we can resume this PR.

**Comment by mtrezza** [comment]
In my understanding, since the FB SDK is a transitive dependency, it should be "invisible" to the developer's app. So if a developer wants to directly call any FB SDK methods, they'd have to separately add the FB SDK as dependency to their app. Moreover, that dependency would be independent from the transitive dependency, in regard to version, compatibility, etc.

Does that sound right? If so, this PR will unlikely be a breaking change and we can merge.

**Comment by L3K0V** [comment]
> `compile` / `api` are long deprecated in favor of `implementation`

That's not correct, only `compile` is deprecated, `api` is used for a transitive dependencies. 

Facebook SDK dependency can be changed from `api` to `implementation` and be used only within the library. In this scenario the developer should take care of adding its own version of the Facebook SDK and use it accordingly to its source code. I'm not aware how the Facebook SDK is configured to work and what will happen if there are two versions - one used by the app and another used by the Parse SDK and if the dependency resolution will cause some troubles.

The point having the Facebook as a `api` dependency I think is to push the user to use the version bundled with the Parse SDK and do the necessary adjustments to their own Facebook code.

Another approach will be to make the Facebook SDK `runtime` or `compileRuntime` dependency and to expect the SDK to be there in the app, but in such case the Parse SDK should take care to cover as many as possible Facebook versions and their breaking changes inside the implementation.

I cannot take a decision here, not until I research how is done in other similar cases and situations.

**Comment by mman** [comment]
> > `compile` / `api` are long deprecated in favor of `implementation`
> 
> That's not correct, only `compile` is deprecated, `api` is used for a transitive dependencies.
> 
> Facebook SDK dependency can be changed from `api` to `implementation` and be used only within the library. In this scenario the developer should take care of adding its own version of the Facebook SDK and use it accordingly to its source code. I'm not aware how the Facebook SDK is configured to work and what will happen if there are two versions - one used by the app and another used by the Parse SDK and if the dependency resolution will cause some troubles.

AFAIK It will cause troubles if two different versions of Facebook SDK will end up being used. I have experienced this myself many times. The Android Studio compiler will eventually flag this as "duplicate class" error since Java can not use two .jar files providing the same classes in isolation. 

This can work in C/Objective-C/Swift world where `static` method are isolated per `.o` file and multiple versions of the same method can this way be linked into the same executable, bit it does not work like this in java world as far as I know.

**Comment by mtrezza** [comment]
> That's not correct, only compile is deprecated, api is used for a transitive dependencies.

Yes, sorry for the confusion, should have phrased that better.

> I'm not aware how the Facebook SDK is configured to work and what will happen if there are two versions - one used by the app and another used by the Parse SDK and if the dependency resolution will cause some troubles.

> AFAIK It will cause troubles if two different versions of Facebook SDK will end up being used. ... The Android Studio compiler will eventually flag this as "duplicate class" error.

There can't be 2 different FB SDK versions used at the same time. Gradle will try to decide on a single version. If it cannot, it will throw an error. If a developer implements their own FB SDK with a higher version, gradle will override the FB SDK version in the Parse SDK. We have to tell gradle that we *did not test* the Parse SDK with any version higher than 12.1.0. We can only assume that it will be compatible with `>=12.1.0 <13.0.0` according to semver. If we don't specify an upper range and only say `>=12.1.0`, we are effectively misleading gradle (and the developer).

If a developer tries to implement a FB SDK that is incompatible with Parse SDK's dependency of the FB SDK, gradle *should* throw an error, that is nothing bad but necessary.

In addition, we are only implementing the FB Login SDK, that means changes are high that a developer adds their own FB SDK to the app, because all other FB modules are missing.

**Comment by L3K0V** [comment]
@mman my point was about the resolution strategy of Gradle which version will pick of many and what will be the behaviour after that. In your experience and from your perspective what should be the approach here?

a) Facebook SDK as `runtime` dependency covering many cases in the Parse SDK
b) Facebook SDK as `api` dependency and the developer to use same version in their app + adjustments on their own FB code
c) Facebook SDK as `implementation` and hoping Gradle to pick up most suitable version between SDK and the app.

Maybe is a manner of how do we want to 'communicate' to developers, do we want to push them to use bundled version, or their own or something else.

**Comment by L3K0V** [comment]
I haven't used before the versions constraints @mtrezza, where you can specify a range. It sounds promising but again maybe we should discuss whenever the Parse SDK bundle the Facebook SDK, bundle + export (`api`) or expect it runtime.

**Comment by mtrezza** [comment]
> my point was about the resolution strategy of Gradle which version will pick of many and what will be the behaviour after that. In your experience and from your perspective what should be the approach here?

That's answered in the [dependency resolution management](https://docs.gradle.org/current/userguide/dependency_management.html#understanding_dependency_resolution) and my previous 2 comments.

> Maybe is a manner of how do we want to 'communicate' to developers, do we want to push them to use bundled version, or their own or something else

Gradle communicates that to the developer. Even if we specify an upper range, the developer can still decide to force override the FB SDK version in the Parse SDK to resolve the resolution conflict. But it should never happen in the background without the developer knowing.  The best way to communicate that is not via a changelog but via gradle.

**Comment by L3K0V** [comment]
> How about we add the range in this PR and open a new issue to discuss the larger question you posed?

Sounds great 👍

**Comment by mtrezza** [comment]
@mman What do you think?

File: facebook/build.gradle
Comment (mtrezza) [inline facebook/build.gradle:38]: Change to 
```
api "com.facebook.android:facebook-login:[13.2.0, 14.0["
```

**Review by mtrezza — CHANGES_REQUESTED** [review]

**Comment by mtrezza** [comment]
@mobilekosmos Would you take a look at the small change requested and the failing CI so we can merge this?

**Comment by mtrezza** [comment]
Does anyone want to continue this PR to get is merged?

**Comment by rommansabbir** [comment]
> Does anyone want to continue this PR to get it merged? It seems to need just a rebase and a trivial code change. It seems that @mobilekosmos has [abandoned](https://github.com/parse-community/Parse-SDK-Android/pull/1163#issuecomment-1107830008) it.

I will try to start again with this PR, once I'm back to my working place.

**Comment by mtrezza** [comment]
Cool, @rommansabbir can I add you to the Parse Android SDK review team? No obligations, but you would get a notification once a PR is ready for review and can do a review to get it merge-ready.

**Comment by rommansabbir** [comment]
> Cool, @rommansabbir can I add you to the Parse Android SDK review team? No obligations, but you would get a notification as member of `@parse-community/android-sdk-review` once a PR is ready for review and can do a review to get it merge-ready.

Yes, that would be amazing!

**Comment by rommansabbir** [comment]
@mtrezza I have updated FB SDK (api) version to a specific range and run all the tests from the existing tests, all passed but I'm unable to push my changes to @mobilekosmos HEAD:master to that the changes could refelect in this PR.

@mobilekosmos can you check, if it's allowed to push commits to this PR?
image

Or, else I just create a new PR and refer to this one? @mman @mtrezza

**Comment by mtrezza** [comment]
New PR please

**Comment by mtrezza** [comment]
Closing, replaced by https://github.com/parse-community/Parse-SDK-Android/pull/1172.

**Pull request closed without merging**
